src/Hadolint/Rules.hs

```diff
@@ -459,33 +461,45 @@ pipVersionPinned = instructionRule code severity message check
     message =
         "Pin versions in pip. Instead of `pip install <package>` use `pip install \
         \<package>==<version>`"
-    check (Run (Arguments args)) =
-        not (isPipInstall args) ||
-        (isRecursiveInstall args || isSetupPyInstall args || all versionFixed (packages args))
+    check (Run (Arguments args)) = not (isPipInstall args) || all versionFixed (packages args)
     check _ = True
-    isVersionedGit :: String -> Bool
+    isPipInstall :: [String] -> Bool
+    isPipInstall cmd =
+        case getInstallArgs cmd of
+            Nothing -> False
+            Just args -> not (["-r"] `isInfixOf` args || ["."] `isInfixOf` args)
+    packages cmd =
+        case getInstallArgs cmd of
+            Nothing -> []
+            Just args -> findPackages args
+    getInstallArgs = stripInstallPrefix isInstallCommand
+    isInstallCommand ('p':'i':'p':_) = True
+    isInstallCommand _ = False
+    versionFixed package = hasVersionSymbol package || isVersionedGit package
     isVersionedGit package = "git+http" `isInfixOf` package && "@" `isInfixOf` package
     versionSymbols = ["==", ">=", "<=", ">", "<", "!="]
-    hasVersionSymbol :: String -> Bool
     hasVersionSymbol package = or [s `isInfixOf` package | s <- versionSymbols]
-    versionFixed :: String -> Bool
-    versionFixed package = hasVersionSymbol package || isVersionedGit package
-    isPipInstall :: [String] -> Bool
-    isPipInstall cmd =
-        ["pip", "install"] `isInfixOf` cmd ||
-        ["pip3", "install"] `isInfixOf` cmd || ["pip2", "install"] `isInfixOf` cmd
-    isRecursiveInstall :: [String] -> Bool
-    isRecursiveInstall cmd = ["-r"] `isInfixOf` cmd
-    isSetupPyInstall :: [String] -> Bool
-    isSetupPyInstall cmd = ["."] `isInfixOf` cmd
-    -- | Returns all the packages after pip install
-    packages :: [String] -> [String]
-    packages args = concat [filter noOption cmd | cmd <- bashCommands args, isPipInstall cmd]
-      where
-        noOption arg = arg `notElem` commandName && not (isFlag arg)
-        commandName = ["pip", "pip2", "pip3", "install"]
-        isFlag ('-':_) = True
-        isFlag _ = False
+
+-- | Returns all the packages after pip install
+findPackages :: [String] -> [String]
+findPackages = takeWhile (not . isEnd) . dropWhile isFlag
+  where
+    isEnd word = isFlag word || word == "&&"
```

Comment:
I suppose `&&` is to stop before next chained command. If that is the case, we should probably also count on others like `;` and `||` or `|`. I believe there will be a use case for them.

Reply:
Yeah, good idea. I'll add `;` and `||`